files/usr/share/cinnamon/applets/menu@cinnamon.org/applet.js

```diff
@@ -3238,37 +3240,29 @@ class CinnamonMenuApplet extends Applet.TextIconApplet {
     }
 
     _listApplications(category_menu_id, pattern){
-        var applist = [];
-        if (category_menu_id) {
-            applist = category_menu_id;
-        } else {
-            applist = "all";
+        if(!pattern) {
+            let res = new Set();
+            res.add(category_menu_id ? category_menu_id : "all");
+            return [res, null];
         }
-        let res;
+        let regexpPattern = new RegExp("\\b"+pattern);
+        let res = new Set();
         let exactMatch = null;
-        if (pattern){
-            res = [];
-            let regexpPattern = new RegExp("\\b"+pattern);
-            for (let i in this._applicationsButtons) {
-                let app = this._applicationsButtons[i].app;
-                let latinisedLowerName = Util.latinise(app.get_name().toLowerCase());
-                if (latinisedLowerName.match(regexpPattern) !== null) {
-                    res.push(app.get_id());
-                    if (!exactMatch && latinisedLowerName === pattern)
-                        exactMatch = app.get_id();
-                }
+        this._applicationsButtons.forEach(button => {
+            let app = button.app;
+            let latinisedLowerName = Util.latinise(app.get_name().toLowerCase());
+            if (latinisedLowerName.match(regexpPattern)) {
+                res.add(app.get_id());
+                if (!exactMatch && latinisedLowerName === pattern)
+                    exactMatch = app.get_id();
             }
-            if (!exactMatch) {
-                for (let i in this._applicationsButtons) {
-                    let app = this._applicationsButtons[i].app;
-                    if (Util.latinise(app.get_name().toLowerCase()).indexOf(pattern)!==-1 ||
-                        (app.get_keywords() && Util.latinise(app.get_keywords().toLowerCase()).indexOf(pattern)!==-1) ||
-                        (app.get_description() && Util.latinise(app.get_description().toLowerCase()).indexOf(pattern)!==-1) ||
-                        (app.get_id() && Util.latinise(app.get_id().slice(0, -8).toLowerCase()).indexOf(pattern)!==-1))
-                        res.push(app.get_id());
-                }
+            if (this.searchInKeywords && app.get_keywords() && Util.latinise(app.get_keywords().toLowerCase()).split(';').some(keyword => keyword.startsWith(pattern))) {
```

Comment:
This could probably be broken down into multiple lines, it's not easy to read this way.